### PR TITLE
Python 3 Port[In progress]

### PR DESCRIPTION
**Pull request by Saumya-Mishra9129**
Trying to port from python 2 to 3 , as Sugar-toolkit is also ported , I am getting this error, get stuck at this.Kindly look at error.
Screenshot from 2020-03-25 00-18-15

and do review also.Thanks

### CODE REVIEW

**Comment by chimosky** [comment]
The activity hasn't been ported to Gtk3, you should do that first. The error you got is as a result of that.
Gtk3 activities use `toolbarbox` and not `toolbox`.

**Comment by quozl** [comment]
@Saumya-Mishra9129, next time please copy and paste errors into GItHub markdown text blocks, using _```_ to delimit.  Images are expensive, and can't be searched.

**Comment by Saumya-Mishra9129** [comment]
@quozl  Okay sure. 
@chimosky  thanks. Port to Gtk3 is still in progress. I am working on it.

**Comment by chimosky** [comment]
> Port to Gtk3 is in progress. I am working on it.

Great, we'll wait for it.

**Comment by Saumya-Mishra9129** [comment]
@pro-panda @quozl @chimosky I have tried to fix toolbar and Vte errors . Please Review. Also some of the methods are removed from Vte.Terminal class in Vte 2.91 as fork_command(), set_emulation() and set_visible_bell(). I couldn't figure out what to do in such cases. Can you suggest here.??

**Comment by chimosky** [comment]
I think `spawn_async` replaced `fork_command`, the [changelog](https://gitlab.gnome.org/GNOME/vte/-/blob/master/ChangeLog.pre-git#L5850) confirms that, one thing you can do is check what those functions do in python2 by importing `vte` in a python2 REPL and find out how to do those things with `Vte.Terminal`.

**Comment by Saumya-Mishra9129** [comment]
@srevinsaju I looked at Terminal activity which also contains Vte. You have worked there , Can you have a look at this activity?

**Comment by srevinsaju** [comment]
@Saumya-Mishra9129 i will check

**Comment by quozl** [comment]
@Saumya-Mishra9129, the Vte library was changed a few times, may I suggest;

* search for the activities that instantiate `Vte.Terminal`,
* look through the commits in those activities for changes in support of the various versions of Vte,
* search older versions of those activities for those removed methods, then search for commits that ported to later Vte.

**Comment by srevinsaju** [comment]
@Saumya-Mishra9129 Please review https://github.com/Saumya-Mishra9129/Frotz/pull/1

Currently the `frotz` binary doesn't work;
This is possibly because some of the needed binaries doesn't work
Please recompile frotz and include all the necessary libs within `lib` to make this work

To include `LD_LIBRARY_PATH`, you can mention it in `envv` args in `spawn_sync`. I have not used `spawn_async`, instead used `spawn_sync`. I hope it would be equally good

**Comment by Saumya-Mishra9129** [comment]
> I have not used spawn_async, instead used spawn_sync. I hope it would be equally good

@srevinsaju I just looked at VTE documentation and found that spawn_sync is deprecated since version 0.48. For your reference -https://lazka.github.io/pgi-docs/Vte-2.91/classes/Terminal.html#Vte.Terminal.spawn_sync

**Comment by srevinsaju** [comment]
@Saumya-Mishra9129 
See https://github.com/Saumya-Mishra9129/Frotz/pull/2

-----------

Although deprecated; its still usable; Anyways, I have used `spawn_async` on PR 2

**Comment by chimosky** [comment]
Reviewed f38db26.

An issue wasn't opened for frotz binaries in this activity but I think you should remove the binaries, the binaries exists for users who would run the activity on an XO laptop; porting to python3 removes the need for that.
- [ ] Change to using system specific package for frotz and specify dependency in README.

**Comment by quozl** [comment]
> * [ ]  Change to using system specific package for frotz and specify dependency in README.

Agreed.  This should also result in a fix for https://github.com/sugarlabs/Frotz/issues/2.

Reminder of standing advice for future use: [_make sure the activity works properly before porting, close any solved issues, merge any pull requests or branches and release the last Python 2 version_](https://github.com/sugarlabs/sugar-docs/blob/master/src/python-porting-guide.md#how-to-port-an-activity-to-python-3).

**Comment by quozl** [comment]
Re: https://github.com/sugarlabs/Frotz/pull/7/commits/d57cfd50f6c1f0b50f225e1dc9a70982045a1a19, my preference is to use `Vte.Terminal.spawn_sync` until it is unavailable.  `Vte.Terminal.spawn_async` is new, and not available, for example, in Ubuntu 18.04.  Moving quickly just because someone shouts **deprecated** at us breaks compatibility with older operating systems without a good reason.

Or, you can use `hasattr` and support both methods.

**Comment by Saumya-Mishra9129** [comment]
> Re: d57cfd5, my preference is to use Vte.Terminal.spawn_sync until it is unavailable. Vte.Terminal.spawn_async is new, and not available, for example, in Ubuntu 18.04. Moving quickly just because someone shouts deprecated at us breaks compatibility with older operating systems without a good reason.

Thanks , I tested spawn_async is not working with Ubuntu 18.04.

> Or, you can use hasattr and support both methods

Yeah It can be best.

**Comment by Saumya-Mishra9129** [comment]
> Change to using system specific package for frotz and specify dependency in README.

I have tried this in b3bc77d and c52c1da. Activity still doesn't work and with no errors in the log , maybe something is broken.

**Comment by Saumya-Mishra9129** [comment]
> I have tried this in b3bc77d and c52c1da. Activity still doesn't work and with no errors in the log , maybe something is broken.

Finally got it working fine. :smile:  @quozl @chimosky @srevinsaju @JuiP Please review.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+What is this?
+=============
+
+Frotz.activity is a simple front end to the classic interactive fiction interpreter frotz on the XO laptop.  Frotz.activity is licensed under the GNU General Public License version 3 (see the file COPYING for details). 
+
+The actual interactive interpreter, Frotz, is licensed under the GNU General Public License version 2.  This xo bundle includes only a binary (version 2.43), but you can obtain the source from here: http://frotz.homeunix.org/frotz/
```
Comment (quozl) [inline README.md]: Says here it only includes a binary, but you removed the binary.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+What is this?
+=============
+
+Frotz.activity is a simple front end to the classic interactive fiction interpreter frotz on the XO laptop.  Frotz.activity is licensed under the GNU General Public License version 3 (see the file COPYING for details).
```
Comment (quozl) [inline README.md]: Let's clearly differentiate between the activity and the upstream software it is a wrapper for.  I suggest changing "Frotz.activity" to "Frotz activity" in this paragraph, and from line 8 onwards when the context is a reference to the activity instead of the Frotz upstream software.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+* [How to use Sugar](https://help.sugarlabs.org/),
+
+Frotz depends on Python, [Sugar
+Toolkit](https://github.com/sugarlabs/sugar-toolkit-gtk3), GTK+ 3, Pango, Vte and frotz.
+
+
+How to install dependencies?
+===========================================
+
+Install the frotz
```
Comment (quozl) [inline README.md]: "the frotz"?  :grin:

File: activity/activity.info [hunk 4]
```diff
@@ -2,8 +2,8 @@
...
 activity_version = 3
 bundle_id = vu.lux.olpc.Frotz
-exec = sugar-activity frotz.FrotzActivity
+exec = sugar-activity3 frotz.FrotzActivity
 icon = activity-frotz
 mime_types = application/x-zmachine
 license = GPLv3+
-host_version = 1
\ No newline at end of file
+host_version = 1
```
Comment (quozl) [inline activity/activity.info]: host version is no longer required.

File: frotz.py [hunk 5]
```diff
@@ -126,15 +131,34 @@ def start_game(self, game_file):
...
-                    self._vte.feed_child("cd '%s'; clear; frotz32|head -3 ; echo '\nLoading %s...'; sleep 2; frotz32 '%s'; exit\n" % (save_dir, os.path.basename(game_file), game_file))
+                dialog = Gtk.MessageDialog(
+                    transient_for=self,
+                    flags=0,
+                    message_type=Gtk.MessageType.INFO,
+                    buttons=Gtk.ButtonsType.OK,
+                    text="frotz is not installed"
+                )
+                dialog.format_secondary_text(
+                    "Please Install frotz , follow README"
```
Comment (quozl) [inline frotz.py]: A space before the comma.  Also the file README is not easily accessible to the child user, so this doesn't help.

File: frotz.py [hunk 6]
```diff
@@ -235,61 +283,63 @@ def _configure_vte(self):
...
+
+        self.set_scrollback_lines(int(scrollback_lines))
         self.set_allow_bold(True)
-        
+
         if conf.has_option('terminal', 'scroll_on_keystroke'):
             scroll_key = conf.getboolean('terminal', 'scroll_on_keystroke')
         else:
-            scroll_key = False
+            scroll_key = 'False'
```
Comment (quozl) [inline frotz.py]: Should be a boolean.  Reference the Terminal activity and PyGObject docs for the function.

File: frotz.py [hunk 7]
```diff
@@ -235,61 +283,63 @@ def _configure_vte(self):
...
             conf.set('terminal', 'scroll_on_keystroke', scroll_key)
         self.set_scroll_on_keystroke(scroll_key)
 
         if conf.has_option('terminal', 'scroll_on_output'):
             scroll_output = conf.getboolean('terminal', 'scroll_on_output')
         else:
-            scroll_output = False
+            scroll_output = 'False'
```
Comment (quozl) [inline frotz.py]: Ditto.

File: frotz.py [hunk 8]
```diff
@@ -235,61 +283,63 @@ def _configure_vte(self):
...
-        
-        self.set_cursor_blinks(blink)
+        blink = Vte.CursorBlinkMode(0) if blink == 'False' else Vte.CursorBlinkMode(1)
+        self.set_cursor_blink_mode(blink)
 
         if conf.has_option('terminal', 'bell'):
             bell = conf.getboolean('terminal', 'bell')
         else:
-            bell = False
+            bell = 'False'
```
Comment (quozl) [inline frotz.py]: Should be a boolean.

File: frotz.py [hunk 9]
```diff
@@ -235,61 +283,63 @@ def _configure_vte(self):
...
         if conf.has_option('terminal', 'scrollback_lines'):
             scrollback_lines = conf.getint('terminal', 'scrollback_lines')
         else:
-            scrollback_lines = 1000
+            scrollback_lines = '1000'
             conf.set('terminal', 'scrollback_lines', scrollback_lines)
-            
-        self.set_scrollback_lines(scrollback_lines)
+
+        self.set_scrollback_lines(int(scrollback_lines))
```
Comment (quozl) [inline frotz.py]: Not sure why this changes.

File: frotz.py [hunk 10]
```diff
@@ -13,94 +13,99 @@
...
+        self._vte.set_encoding('utf-8')
         self._vte.show()
         self._vte.connect("child-exited", self._quit_cb)
 
-        scrollbar = gtk.VScrollbar(self._vte.get_adjustment())
+        scrollbar = Gtk.VScrollbar(self._vte.get_vadjustment())
         scrollbar.show()
 
-        box.pack_start(self._vte)
+        box.pack_start(self._vte, True, True, 0)
```
Comment (srevinsaju) [inline frotz.py]: providing `kwargs` might make the code more readable

File: frotz.py [hunk 11]
```diff
@@ -126,15 +131,34 @@ def start_game(self, game_file):
             # print a welcome banner and pause for a moment
             # that way the end user will have a chance to see which version of frotz we are using
             # and which file we are loading
-
-            if platform.machine().startswith('arm'):
-                logging.error('ARM not supported, yet') # FIXME
-                sys.exit(0)
+            cmd = '/usr/games/frotz'
```
Comment (srevinsaju) [inline frotz.py]: Are you sure `frotz` is always installed in `/usr/games/frotz` across distros?

File: frotz.py [hunk 12]
```diff
@@ -143,19 +167,19 @@ def read_file(self, file_path):
...
         tmpfolder = mkdtemp('.tmp', 'url', os.path.join(self.get_activity_root(), 'instance'))
         tmpfilepath = os.path.join(tmpfolder, 'url')
         try:
             tmpfile = open(tmpfilepath, 'w')
             tmpfile.write(url)
             tmpfile.close()
-            os.chmod(tmpfolder, 0755)
-            os.chmod(tmpfilepath, 0755)
+            os.chmod(tmpfolder, 0o755)
+            os.chmod(tmpfilepath, 0o755)
```
Comment (srevinsaju) [inline frotz.py]: out of curiosity and due to security reasons, maybe we should not make a temp file executable for the world?

File: frotz.py [hunk 13]
```diff
@@ -187,30 +211,54 @@ def _paste_cb(self, button):
...
-        self.fork_command()
+class VTE(Vte.Terminal):
 
-    def _configure_vte(self):
-        conf = ConfigParser.ConfigParser()
+    def __init__(self):
+        Vte.Terminal.__init__(self)
+        self.configure_terminal()
+
+        os.chdir(os.environ["HOME"])
```
Comment (srevinsaju) [inline frotz.py]: can you provide an inline comment explaining why this was uncommented or why this is necessary?

File: setup.py [hunk 14]
```diff
@@ -1,4 +1,4 @@
 #!/usr/bin/python
```
Comment (srevinsaju) [inline setup.py]: It would be better to use `/usr/bin/env python3`, as `python` does not exist on Ubuntu 20.04

**Review by srevinsaju — CHANGES_REQUESTED** [review]

File: activity/activity.info (hunk 4 above)
Comment (Saumya-Mishra9129) [reply activity/activity.info]: Removed

File: README.md (hunk 3 above)
Comment (Saumya-Mishra9129) [reply README.md]: :sweat_smile:

File: README.md (hunk 1 above)
Comment (Saumya-Mishra9129) [reply README.md]: Yeah Done

File: README.md (hunk 2 above)
Comment (Saumya-Mishra9129) [reply README.md]: Thanks changed!

File: frotz.py (hunk 11 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Yeah , I have checked on Debian and Ubuntu, Not sure about fedora. Do you use Fedora, Can you check once?

File: frotz.py (hunk 12 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Yeah I was also curious to know why they are doing this. Isn't there any other way? :thinking:

File: frotz.py (hunk 9 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Because of `TypeError: Must be number, not str`, scrollback_lines takes int as argument , not str.

File: frotz.py (hunk 8 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Because` configparser accepts <class 'str'> as *args.`

File: frotz.py (hunk 6 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Because , configparser accepts <class 'str'> as *args.

Comment (Saumya-Mishra9129) [reply frotz.py]: @quozl See https://docs.python.org/3/library/configparser.html#configparser.ConfigParser.set, I wonder about terminal activity, do we need changes to terminal activity.?

File: setup.py (hunk 14 above)
Comment (Saumya-Mishra9129) [reply setup.py]: Done!

File: frotz.py (hunk 10 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Done

File: frotz.py (hunk 13 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Thanks for noticing, The line seems to have no effect. Actually activity was crashing before, so I copied some of the code from VncLauncher activity. I think it is good to remove line, if it has no effect.

File: frotz.py (hunk 5 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Thanks @quozl , will it work this way c388c5c ?

File: frotz.py (hunk 11 above)
Comment (srevinsaju) [reply frotz.py]: Fedora saves it to the /usr/bin as other distributions do. Only Linux distros that use `apt` saves it to `/usr/games`. All others, Fedora, RHEL, CentOS, Arch Linux and Manjaro installs it to `/usr/bin`

File: frotz.py (hunk 12 above)
Comment (srevinsaju) [reply frotz.py]: Change it to 644 perhaps

File: frotz.py (hunk 13 above)
Comment (srevinsaju) [reply frotz.py]: Great

File: frotz.py (hunk 11 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Thanks Done in 5c38b6c

File: frotz.py (hunk 12 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Great , done in 607dd6f

File: frotz.py [hunk 15]
```diff
@@ -131,37 +131,34 @@ def start_game(self, game_file):
...
+                            "cd '%s'; "
+                            "clear; "
+                            "frotz64|head -3 ; "
+                            "echo '\nLoading %s...'; "
+                            "sleep 2; frotz '%s'; "
+                            "exit\n" % (
+                                save_dir,
+                                os.path.basename(game_file),
+                                game_file)).encode('utf-8')
+                )
```
Comment (chimosky) [inline frotz.py]: Why this indentation?

File: README.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
+What is this?
+=============
+
+Frotz activity is a simple front end to the classic interactive fiction interpreter Frotz on the XO laptop.  Frotz activity is licensed under the GNU General Public License version 3 (see the file COPYING for details).
```
Comment (quozl) [inline README.md]: No longer for the XO laptop.

Comment (quozl) [reply README.md]: Details of the licensing are actually in frotz.py, not in COPYING.  COPYING does contain details, but not the copyright and license.  If in doubt as to the best thing to say; what is in frotz.py header is exactly what should be said.

File: README.md [hunk 17]
```diff
@@ -0,0 +1,31 @@
+What is this?
+=============
+
+Frotz activity is a simple front end to the classic interactive fiction interpreter Frotz on the XO laptop.  Frotz activity is licensed under the GNU General Public License version 3 (see the file COPYING for details). 
+
+The actual interactive interpreter, Frotz, is licensed under the GNU General Public License version 2. You can obtain the source from here: http://frotz.homeunix.org/frotz/
```
Comment (quozl) [inline README.md]: Link does not work.

Comment (quozl) [reply README.md]: License of Frotz includes the _or later version_ wording.

File: README.md [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
+* [How to use Sugar](https://help.sugarlabs.org/),
+
+Frotz activity depends on Python, [Sugar
+Toolkit](https://github.com/sugarlabs/sugar-toolkit-gtk3), GTK+ 3, Pango, Vte and frotz.
+
+
+How to install dependencies?
+===========================================
```
Comment (quozl) [inline README.md]: Underline extends beyond heading.

File: frotz.py [hunk 19]
```diff
@@ -112,10 +117,10 @@ def __init__(self, handle):
...
         # when we return to the idle state, launch the default game
         # if read_file is called, that will override this
-        gobject.idle_add(self.start_game, default_game_file)
+        GLib.idle_add(self.start_game, default_game_file)
     
-    def _quit_cb(self, foo=None):
-        print "Quitting..."
+    def _quit_cb(self, vte, foo=None):
+        logging.debug("Quitting...")
         sys.exit(0)
```
Comment (quozl) [inline frotz.py]: Use Activity.close instead?  See what Terminal does; when last tab is clsoed it calls `self.close()`.

File: frotz.py [hunk 20]
```diff
@@ -126,15 +131,45 @@ def start_game(self, game_file):
             # print a welcome banner and pause for a moment
             # that way the end user will have a chance to see which version of frotz we are using
             # and which file we are loading
-
-            if platform.machine().startswith('arm'):
-                logging.error('ARM not supported, yet') # FIXME
-                sys.exit(0)
+            if platform.version().find("Ubuntu") > -1 or platform.version().find("Debian") > -1:
```
Comment (quozl) [inline frotz.py:180]: While this works for Ubuntu and Debian, a more correct method is to search for the binary in a list of paths, like how Linux does it.

File: frotz.py [hunk 21]
```diff
@@ -126,15 +131,45 @@ def start_game(self, game_file):
...
             else:
-                if platform.architecture()[0] == '64bit':
-                    self._vte.feed_child("cd '%s'; clear; frotz64|head -3 ; echo '\nLoading %s...'; sleep 2; frotz64 '%s'; exit\n" % (save_dir, os.path.basename(game_file), game_file))
-                else:
-                    self._vte.feed_child("cd '%s'; clear; frotz32|head -3 ; echo '\nLoading %s...'; sleep 2; frotz32 '%s'; exit\n" % (save_dir, os.path.basename(game_file), game_file))
+                cmd = '/usr/bin/frotz'
+            if os.path.isfile(cmd) and os.access(cmd, os.X_OK):
+                logging.debug('Using frotz installed in the system')
+                self._vte.feed_child(
+                    (
```
Comment (quozl) [inline frotz.py]: Indentation here is very strange; the open parenthesis is not aligned with the close.

File: frotz.py [hunk 22]
```diff
@@ -126,15 +131,45 @@ def start_game(self, game_file):
...
+            else:
+                dialog = Gtk.MessageDialog(
+                    transient_for=self,
+                    flags=0,
+                    message_type=Gtk.MessageType.INFO,
+                    buttons=Gtk.ButtonsType.OK,
+                    text="frotz is not installed"
+                )
+                dialog.format_secondary_text(
+                    "Install frotz by clicking OK and then restart the activity"
```
Comment (quozl) [inline frotz.py]: It should not be necessary to restart the activity.

File: frotz.py [hunk 23]
```diff
@@ -126,15 +131,45 @@ def start_game(self, game_file):
...
+                )
+                response = dialog.run()
+
+                if response == Gtk.ResponseType.OK:
+                    logging.debug("Installing frotz")
+                    if platform.version().find("Ubuntu") > -1 or platform.version().find("Debian") > -1:
+                        cmd = "apt install frotz"
```
Comment (quozl) [inline frotz.py]: Doesn't the user have to be root in order to use these commands?  We don't run Sugar activities as root.

File: frotz.py [hunk 24]
```diff
@@ -143,19 +178,19 @@ def read_file(self, file_path):
...
         tmpfolder = mkdtemp('.tmp', 'url', os.path.join(self.get_activity_root(), 'instance'))
         tmpfilepath = os.path.join(tmpfolder, 'url')
         try:
             tmpfile = open(tmpfilepath, 'w')
             tmpfile.write(url)
             tmpfile.close()
-            os.chmod(tmpfolder, 0755)
-            os.chmod(tmpfilepath, 0755)
+            os.chmod(tmpfolder, 0o755)
+            os.chmod(tmpfilepath, 0o644)
```
Comment (quozl) [inline frotz.py]: https://github.com/sugarlabs/Frotz/pull/7/commits/607dd6ff8a5704fbe14a960443de12e6bddf9721 doesn't explain in the commit message _why_ you are changing the file mask.

File: frotz.py [hunk 25]
```diff
@@ -187,30 +222,54 @@ def _paste_cb(self, button):
...
+            self.spawn_sync(
+                Vte.PtyFlags.DEFAULT,
+                os.environ["HOME"],
+                ["/bin/bash"],
+                [],
+                GLib.SpawnFlags.DO_NOT_REAP_CHILD,
+                None,
+                None)
+        else:
+            self.fork_command_full(
```
Comment (quozl) [inline frotz.py:254]: Why is it necessary to support versions of Vte without spawn_sync?

File: frotz.py (hunk 8 above)
Comment (quozl) [reply frotz.py]: Not a good enough reason.  What matters here is what Vte requires.

File: frotz.py (hunk 9 above)
Comment (quozl) [reply frotz.py]: Change line 319 back to integer, not string.

File: frotz.py (hunk 6 above)
Comment (quozl) [reply frotz.py]: I've seen no such problem on Terminal.

**Review by quozl — COMMENTED** [review]
Thanks.  Reviewed again.

**Comment by quozl** [comment]
Tested on Ubuntu 20.04;

* [ ] the stop button is in the wrong place,
* [ ] the dialog is not standard for Sugar, use an Alert class,
* [ ] clicking on "Get more games" just leads me to a Journal object for a link; it should launch Browse, (see Terminal for how),
* [ ] the more games link is external (at OLPC); it should be hosted in the repository,
* [ ] the install prepares a command but does not execute it,
* [ ] the install does not work, Permission denied,
* [ ] the font is very small, and is not the font configured by Sugar,
* [ ] there's no way to change the font size in the activity,

File: frotz.py (hunk 11 above)
Comment (quozl) [reply frotz.py]: /usr/games is in $PATH on Debian and Ubuntu, so use $PATH.

File: frotz.py (hunk 9 above)
Comment (quozl) [reply frotz.py]: Not a good enough reason to change, in my opinion.

Comment (Saumya-Mishra9129) [reply frotz.py]: @quozl https://github.com/sugarlabs/Frotz/pull/7/commits/813abae39151c64aebca002642f97413fc204c07 This should be fine.

File: frotz.py (hunk 8 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: https://github.com/sugarlabs/Frotz/pull/7/commits/813abae39151c64aebca002642f97413fc204c07 Thanks Please review.

File: frotz.py (hunk 25 above)
Comment (Saumya-Mishra9129) [reply frotz.py:254]: Actually I took terminal activity as reference. Terminal uses fork_command_full and spawn_sync https://github.com/sugarlabs/terminal-activity/blob/master/terminal.py#L496.

File: README.md (hunk 18 above)
Comment (Saumya-Mishra9129) [reply README.md]: Done in c89585f

File: README.md (hunk 17 above)
Comment (Saumya-Mishra9129) [reply README.md]: Done in c89585f

File: README.md (hunk 16 above)
Comment (Saumya-Mishra9129) [reply README.md]: Done in c89585f

File: frotz.py (hunk 19 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Done in 298e911

File: frotz.py (hunk 24 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: https://github.com/sugarlabs/Frotz/pull/7/commits/69f178133e99c6ec2001a187df50a027eb8e997a Please see

File: frotz.py (hunk 21 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Done in 480c207

File: frotz.py (hunk 15 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Done in 480c207.

**Comment by Saumya-Mishra9129** [comment]
> the more games link is external (at OLPC); it should be hosted in the repository,

@quozl What do you mean by it? I couldn't get completely , Is it mean including it in README.?

**Comment by quozl** [comment]
Thanks.

I've tested 9dcd6b3.  Changing font size confuses the output.  I've reproduced without this activity; it happens when the virtual terminal changes size as consequence of font size changes.  I've [asked upstream](https://gitlab.com/DavidGriffith/frotz/-/issues/220).  For the moment, you can remove the font size changes.

Please use sugar3.graphics.style to set the FONT_SIZE.  Finance, Maze, Poll, and Words all use the Sugar setting.

> > the more games link is external (at OLPC); it should be hosted in the repository,
> 
> @quozl What do you mean by it? I couldn't get completely , Is it mean including it in README.?

At the moment the link is at wiki.laptop.org, which was set up in 2008 when the activity was first written.  In moving the activity to Sugar Labs all relevant sources and information should be in Sugar Labs infrastructure if possible.  GitHub holds the README.md file for us.

Regarding the Vte settings; parts of the activity were copied from Terminal as of 2008, probably so that settings changes made to Terminal would apply to Frotz as well.  I don't think all the settings are useful.  Remove as much as you can.

File: frotz.py [hunk 26]
```diff
@@ -139,15 +139,17 @@ def start_game(self, game_file):
...
+                        "cd '%s'; "
+                        "clear; "
+                        "frotz|head -3 ; "
+                        "echo '\nLoading %s...'; "
+                        "sleep 2; frotz '%s'; "
+                        "exit\n" % (
+                            save_dir,
+                            os.path.basename(game_file),
+                            game_file)
+
```
Comment (chimosky) [inline frotz.py]: Why the empty line?

**Review by chimosky — CHANGES_REQUESTED** [review]

File: frotz.py [hunk 27]
```diff
@@ -125,6 +134,37 @@ def __init__(self, handle):
...
+        zoom_in_button = ToolButton('zoom-in')
+        zoom_in_button.set_tooltip(_('Zoom in'))
+        zoom_in_button.props.accelerator = '<Ctrl>plus'
+        zoom_in_button.connect('clicked', self.__zoom_in_cb)
+        view_toolbar.insert(zoom_in_button, -1)
+        zoom_in_button.show()
+        return view_toolbar
+
+    def _zoom(self, step):
+
```
Comment (chimosky) [inline frotz.py]: Why the empty line?

**Review by chimosky — CHANGES_REQUESTED** [review]

File: frotz.py (hunk 27 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Thanks now I have removed this commit. Please See https://github.com/sugarlabs/Frotz/pull/7#issuecomment-675881114

File: frotz.py (hunk 26 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Removed in c7558d0

**Comment by Saumya-Mishra9129** [comment]
@quozl Please review till 1fe9fbb.

**Comment by quozl** [comment]
Upstream suggested Fizmo.

**Comment by quozl** [comment]
Am reviewing.  Standby.

**Comment by quozl** [comment]
* [ ] the start button is misleading; it does nothing except in the scenario where frotz is not installed; rather than use feed child to send install commands to the terminal, use a subprocess, or packagekit.

**Comment by Saumya-Mishra9129** [comment]
> the start button is misleading; it does nothing except in the scenario where frotz is not installed; rather than use feed child to send install commands to the terminal, use a subprocess, or packagekit.

Thanks @quozl It is fixed in 52f914c.

**Comment by Saumya-Mishra9129** [comment]
> clicking on "Get more games" just leads me to a Journal object for a link; it should launch Browse, (see Terminal for how),

Only this needs a fix.

**Comment by Saumya-Mishra9129** [comment]
a164e54 It does not work this error in log:

```Traceback (most recent call last):
  File "/usr/share/sugar/activities/frotz/frotz.py", line 247, in _get_games_cb
    subprocess.Popen(cmd, shell=False)
  File "/usr/lib/python3.7/subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.7/subprocess.py", line 1522, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'xdg-open http://wiki.laptop.org/go/Frotz/Games': 'xdg-open http://wiki.laptop.org/go/Frotz/Games'
```

**Comment by quozl** [comment]
I suggest using launch_bundle as used by sugarterm.py in Terminal activity.

**Comment by Saumya-Mishra9129** [comment]
> I suggest using launch_bundle as used by sugarterm.py in Terminal activity.

Thanks @quozl Done in ca59b68.:smile:

File: frotz.py [hunk 28]
```diff
@@ -126,58 +138,97 @@ def start_game(self, game_file):
...
             # that way the end user will have a chance to see which version of frotz we are using
             # and which file we are loading
-
-            if platform.machine().startswith('arm'):
-                logging.error('ARM not supported, yet') # FIXME
-                sys.exit(0)
+            if platform.version().find("Ubuntu") > -1 or platform.version().find("Debian") > -1:
+                cmd = '/usr/games/frotz'
+            else:
+                cmd = '/usr/bin/frotz'
```
Comment (srevinsaju) [inline frotz.py]: This still may not work. If I build frotz manually, it always installs to `/usr/local/bin` or `/usr/local/games` (on `apt` based distros). A suggested fix is to do this, as @quozl suggested is to:

https://github.com/sugarlabs-appstore/aslo-v4/blob/41383de6fa580dff1a6863c5d1505d01d5d25050/aslo4/platform/__init__.py#L29-L48
```python
def get_executable_path(executable, raise_error=True):
    """
    Returns the absolute path of the executable
    if it is found on the PATH,
    if it is not found raises a FileNotFoundError
    :param executable:
    :return:
    """

    for i in PATH:
        if os.path.exists(os.path.join(i, executable)):
            return os.path.join(i, executable)
    if raise_error:
        raise FileNotFoundError(
            "Could not find {p} on PATH. "
            "Make sure {p} is added to path and try again".format(
                p=executable))
    else:
        return False
```

Link to original source provided above the code snippet. 
You can do a `try...catch` on `FileNotFoundError` or pass `raise_error=False` and check if the returned value is false to check if Frotz exists on path.

**Review by srevinsaju — CHANGES_REQUESTED** [review]

File: frotz.py (hunk 28 above)
Comment (Saumya-Mishra9129) [reply frotz.py]: Thanks , Tried it in 65c66e6. Tested on debian and Ubuntu works fine. :smiley:

Comment (srevinsaju) [reply frotz.py]: Perfect! :tada:

**Comment by quozl** [comment]
Thanks.  Reviewed.  Pushed review changes.  Fixes #4.  Fixes #3.  Fixes #2.

**Comment by Saumya-Mishra9129** [comment]
Thanks Reviewed till 3a9baf7. Please merge these changes.

**Review by JuiP — APPROVED** [review]
